pkg/policy/mapstate.go

```diff
@@ -114,55 +114,92 @@ func (e *MapStateEntry) IsRedirectEntry() bool {
 	return e.ProxyPort != 0
 }
 
-// Equal returns true of two entries are equal
-func (e *MapStateEntry) Equal(o *MapStateEntry) bool {
+// DatapathEqual returns true of two entries are equal in the datapath's PoV,
+// i.e., both Deny and ProxyPort are the same for both entries.
+func (e *MapStateEntry) DatapathEqual(o *MapStateEntry) bool {
 	if e == nil || o == nil {
 		return e == o
 	}
 
-	return e.IsDeny == o.IsDeny && e.ProxyPort == o.ProxyPort && e.DerivedFromRules.Equals(o.DerivedFromRules)
+	return e.IsDeny == o.IsDeny && e.ProxyPort == o.ProxyPort
 }
 
 // DenyPreferredInsert inserts a key and entry into the map by given preference
 // to deny entries, and L3-only deny entries over L3-L4 allows.
-func (keys MapState) DenyPreferredInsert(key Key, entry MapStateEntry) {
-	// If it is a deny policy, it will take precedence over any other key.
-	if entry.IsDeny {
-		// If there is an allow or a deny L3-L4 and the new deny is a L3-Only,
-		// the new deny takes precedence over all L3-L4 for each the L3 is the
-		// same on both the existing rules and new deny rule.
-		if key.DestPort == 0 {
+func (keys MapState) DenyPreferredInsert(newKey Key, newEntry MapStateEntry) {
+	if newEntry.IsDeny {
+		// case for an existing allow L4-only and we are inserting deny L3-only
+		if newKey.DestPort == 0 && newKey.Nexthdr == 0 && newKey.Identity != 0 {
+			l4OnlyAllows := MapState{}
+			for k, v := range keys {
+				if newKey.TrafficDirection == k.TrafficDirection &&
+					!v.IsDeny &&
+					k.Identity == 0 {
+					// create a deny L3-L4 with the same allowed L4 port and proto
+					newKeyCpy := newKey
+					newKeyCpy.DestPort = k.DestPort
+					newKeyCpy.Nexthdr = k.Nexthdr
+					keys[newKeyCpy] = newEntry
+
+					l4OnlyAllows[k] = v
+				}
+			}
+			// Delete all L3-L4 if we are inserting a deny L3-only and
+			// there aren't allow L4-only for the existing deny L3-L4
 			for k := range keys {
-				if k.TrafficDirection == key.TrafficDirection &&
-					k.Identity == key.Identity {
-					delete(keys, k)
+				if k.TrafficDirection == newKey.TrafficDirection &&
+					k.DestPort != 0 && k.Nexthdr != 0 &&
+					k.Identity == newKey.Identity {
+
+					kCpy := k
+					kCpy.Identity = 0
+					if _, ok := l4OnlyAllows[kCpy]; !ok {
+						delete(keys, k)
+					}
 				}
 			}
+		} else {
+			// Do not insert 'newKey' if the map state already denies traffic
+			// which is a superset of (or equal to) 'newKey'
+			newKeyCpy := newKey
+			newKeyCpy.DestPort = 0
+			newKeyCpy.Nexthdr = 0
+			v, ok := keys[newKeyCpy]
+			if ok && v.IsDeny {
+				// Found a L3-only Deny so we won't accept any L3-L4 policies
+				return
+			}
 		}
-		keys[key] = entry
-		return
-	}
 
-	if _, ok := keys[key]; ok {
-		// Key already exist, keep the existing entry so that
-		// a deny entry is never overwritten by an allow entry
+		keys[newKey] = newEntry
+		return
+	} else if newKey.Identity == 0 {
+		// case for an existing deny L3-only and we are inserting allow L4-only
```

Comment:
Here we could also be inserting a wildcard entry (allow all on all ports). Maybe should add
```
&& newKey.DestPort != 0
```